[UnifiedPDF] [iOS] Web process hangs the first time a pointing device is hovered over text annotations

aprotyas opened the pull request:
#### 2809f525060e8d484414a5caf7b03e7c87673df1
<pre>
[UnifiedPDF] [iOS] Web process hangs the first time a pointing device is hovered over text annotations
<a href="https://bugs.webkit.org/show_bug.cgi?id=288875">https://bugs.webkit.org/show_bug.cgi?id=288875</a>
<a href="https://rdar.apple.com/145882906">rdar://145882906</a>

Reviewed by Aditya Keerthi and Wenson Hsieh.

The first time a pointing device is hovered over a text annotation, we
seem to hit UIScreen initialization code from the web process since we
try to read UIColor.systemBlueColor (to paint hover state on said
annotations). This causes the WP to hang for a while (see
<a href="https://rdar.apple.com/74127113">rdar://74127113</a>), because of existing sandbox restrictions.

We fix this bug by simply not painting hover state for text annotations
on iOS. This is wholly unnecessary anyway and may even give the user the
wrong impression, given a user cannot modify a text annotation in any
manner yet.

Furthermore, we also introduce a more principled behavior change by not
reading UIColor.systemBlueColor directly from the WP anymore. Instead,
we delegate system color processing to RenderTheme, and simply call into
RenderTheme::systemColor(CSSValueAppleSystemBlue, ...). This ensures
that when UnifiedPDFPlugin fully supports text annotations on iOS, we
will not face this same problem again.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paintPDFContent):
(WebKit::UnifiedPDFPlugin::paintPDFSelection):
(WebKit::textAnnotationHoverColor):
(WebKit::UnifiedPDFPlugin::paintHoveredAnnotationOnPage):

Canonical link: <a href="https://commits.webkit.org/291410@main">https://commits.webkit.org/291410@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eb40992edb4e9edc3544c59f863de4f03e17677e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92871 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/12422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2068 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/43397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12702 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20874 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/97867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/43397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95873 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/9572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/84010 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/9264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/42710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/1616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/99889 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/19923 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/99889 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/20174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79906 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/99889 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14831 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/19907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/25083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/19594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/23054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21335 "Build is in progress. Recent messages:") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -718,6 +718,12 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
+    static constexpr bool hasFullAnnotationSupport = false;
+#else
+    static constexpr bool hasFullAnnotationSupport = false;
+#endif
```
Intentional that these are the same?

aprotyas replied:
🤦

aprotyas replied:
Typed too quickly

aprotyas replied:
the PLATFORM(MAC) case should be true.

whsieh commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -718,6 +718,12 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
+#if PLATFORM(MAC)
+    static constexpr bool hasFullAnnotationSupport = false;
+#else
+    static constexpr bool hasFullAnnotationSupport = false;
```
Is one of these incorrect? (Or should the platform define be removed)

aprotyas replied:
Indeed, PLATFORM(MAC) case is incorrect

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -962,11 +964,25 @@ static String mutationObserverNotificationString()
+        if (renderer)
+            styleColorOptions = renderer->styleColorOptions();
+
+        return renderTheme.systemColor(CSSValueAppleSystemBlue, styleColorOptions).colorWithAlpha(textAnnotationHoverAlpha);
```
You can always use the singleton.

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -962,11 +964,25 @@ static String mutationObserverNotificationString()
+
+    static NeverDestroyed color = [element = pluginElement] -> Color {
+        if (!element)
+            return SRGBA<uint8_t> { 0, 122, 255 };
```
Also don't need this, since you can always use the singleton.

aprotyas replied:
Fixed

aprotyas committed:
[UnifiedPDF] [iOS] Web process hangs the first time a pointing device is hovered over text annotations
https://bugs.webkit.org/show_bug.cgi?id=288875
rdar://145882906

Reviewed by Aditya Keerthi and Wenson Hsieh.

The first time a pointing device is hovered over a text annotation, we
seem to hit UIScreen initialization code from the web process since we
try to read UIColor.systemBlueColor (to paint hover state on said
annotations). This causes the WP to hang for a while (see
rdar://74127113), because of existing sandbox restrictions.

We fix this bug by simply not painting hover state for text annotations
on iOS. This is wholly unnecessary anyway and may even give the user the
wrong impression, given a user cannot modify a text annotation in any
manner yet.

Furthermore, we also introduce a more principled behavior change by not
reading UIColor.systemBlueColor directly from the WP anymore. Instead,
we delegate system color processing to RenderTheme, and simply call into
RenderTheme::systemColor(CSSValueAppleSystemBlue, ...). This ensures
that when UnifiedPDFPlugin fully supports text annotations on iOS, we
will not face this same problem again.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paintPDFContent):
(WebKit::UnifiedPDFPlugin::paintPDFSelection):
(WebKit::textAnnotationHoverColor):
(WebKit::UnifiedPDFPlugin::paintHoveredAnnotationOnPage):

Canonical link: https://commits.webkit.org/291410@main
